### DISCUSSION TITLE
How to install the new Sparrow extended UI?

### DISCUSSION THREAD

**Discussion by mriscoc** [mriscoc/Sparrow-Extended-UI-for-RIGOL-DHO800_DHO900#4] [Q&A] [answered]
> [!IMPORTANT]  
> **Make a backup** of your internal µSD card just in case.

The Sparrow Extended application constitutes the graphical user interface (GUI) of the DHO 800/900. It is the primary application of the oscilloscope and provides the user with the full complement of its functionality.

The oscilloscope utilises an Android operating system, thereby enabling the installation of supplementary tools such as the DHOActions panel. This panel facilitates the execution of SCPI scripts or other applications from the main screen.

Luckily, the GUI can be easily replaced by uninstalling the original application, Sparrow, and installing a replacement application with unlocked and extended functionalities. You can easily undo this process at any time by uninstalling the extended version and installing the original from your backup. No other files are changed in the system.

There are two versions of the Sparrow Extended: the "normal" version and the "system" version. The original application from the manufacturer contains special signatures, which prevent us from installing an application with the same privileges. Consequently, after uninstalling the manufacturer's application, it is possible to install the "normal" version, which contains a different signature but fewer privileges. This means that it will not be possible to perform certain operations that require system privileges, such as taking a screenshot from the application.

However, it is possible to install a "system" version, provided that the device has first been patched to accept other signatures than those issued by the manufacturer. In such a case, the application will have all the privileges of the original application. 

> [!WARNING] 
> If you are planning to install the system version, you MUST FIRST obtain system permission by patching the system before installing the system APK. This is an advanced procedure that must be carried out carefully. Look for the guide here: https://github.com/mriscoc/RIGOL_DHO800_DHO900_GUI/discussions/8

For install the Sparrow extended application follow the next steps:

Download and extract the Sparrow extended application to a known folder and open a terminal window ([CMD](https://www.google.com/search?q=how+to+open+a+cmd+window+inside+of+a+folder)) in that folder.  Rename the selected version (normal or system) to **Sparrow.apk**

The [ADB](https://www.google.com/search?q=how+to+install+adb) tool can be obtained from the following link: https://developer.android.com/tools/releases/platform-tools or by uncompressing the file minimal_adb_fastboot where you will find a minimal ADB toolkit, then follow this guide:

```shell
::####################################################################################
::# CONNECT to the Rigol DHO, replace ###.###.###.### with the IP of the device.
::####################################################################################
adb connect ###.###.###.###:55555
adb root

::####################################################################################
::# UNINSTALL old version
::####################################################################################
adb shell am force-stop com.rigol.launcher
adb shell am force-stop com.rigol.scope
adb shell pm clear com.rigol.scope
adb shell pm uninstall com.rigol.scope
adb shell pm uninstall --user 0 com.rigol.scope

::####################################################################################
::# INSTALL the new Sparrow extended
::####################################################################################
adb install -g -r "Sparrow.apk"
adb reboot
```

If you got an error message while trying to install the new Sparrow.apk
```
[INSTALL_FAILED_UPDATE_INCOMPATIBLE: Package com.rigol.scope signatures do not match the previously installed version; ignoring!]
```

Try this procedure and then try to install the new APK again:

```shell
::####################################################################################
::# RENAME Sparrow stock system app and REMOVE the Sparrow data:
::####################################################################################
adb shell
su
mount -o rw,remount /system
mv /system/app/Sparrow/Sparrow.apk /system/app/Sparrow/Sparrow.apk.disabled
rm -rf /data/app/com.rigol.scope-1
sync
reboot

**Comment by Evariste28** [comment]
Hello,
La date et l'heure sont toujours en GMT de la Chine, le GMT de Bruxelles ne reste pas mémorisé.
Version du Sparrow  "Sparrow_Extended_v0.5.1"
Merci par avance, pour votre réponse.

**Reply by mriscoc** [reply]
You need to remove or comment out the line:
`setprop persist.sys.timezone Asia/Shanghai` from the file rigol/shell/start_rigol_app.sh
Then use the settings icon in the app main menu to select the proper timezone.

**Reply by Evariste28** [reply]
Hello, merci pour l'information, j'ai fait la modification, "avec TotalCommander plugin ADB"  tout fonctionne très bien
avec la ligne  "setprop persist.sys.timezone Europe centrale/Bruxelles".
Encore merci. 👍👌

**Reply by Philipp-3** [reply]
Hello,
today I installed your extended ui and set system permission. Worked fine so far, afterwards I tried the following:
`adb connect ###.###.###.###:55555`
`adb root`
`adb pull rigol/shell/start_rigol_app.sh start_rigol_app.sh`
Then I commented out the setprop... line via # locally and saved the file
`adb push start_rigol_app.sh rigol/shell/start_rigol_app.sh`
`adb reboot`

Now my oszi stucks ar the big R image when booting. ADB connect is still possible. Could you help me please?

**Reply by mriscoc** [reply]
That happens because missing some permissions or because code formatting error due to the text editor.

**Reply by Philipp-3** [reply]
Ok, what would be the correct way to comment out the line, and how could I fix it?

**Reply by mriscoc** [reply]
The stock permissions for those scripts are 777

```shell
> adb shell
$ ls -l /rigol/shell/start_rigol_app.sh
-rwxrwxrwx 1 root root 5839 2024-09-07 00:08 /rigol/shell/start_rigol_app.sh
```
So you can use **chmod** to restore those permissions

**Reply by Philipp-3** [reply]
Now it works, thank you for your support!

**Comment by zalexua** [comment]
Is there a way to revert this custom interface back?
What is the way? Reinstall firmware? Could you describe that?

**Reply by mriscoc** [reply]
Yes, uninstall com.rigol.scope, then reinstall your backup copy in the same way that you install the custom apk.

If you don't have a backup, you can get the original application by uncompressing the curren firmware package.

**Comment by mriscoc** [comment, marked as answer]
Updated to include an additional procedure if you want to enable system permissions to take screenshots from within the application

**Comment by SlCKB0Y** [comment]
If someone updates their scope type from dho802 to dho824 to get the increase in specs using a modified vendor.bin, is there a way to adjust the interface so that the two extra channels which don't physically exist dont show up? that would be awesome if that could be user configured.

**Reply by mriscoc** [reply]
This could be done, but it will only be a cosmetic solution for the main screen, all other options will list the additional channels.

**Reply by mriscoc** [reply]
The new version will unlock all the functionality for the DHO802 without change the device model.

**Reply by mriscoc** [reply]
![DHO802_about](https://github.com/user-attachments/assets/c37119f9-13bf-4b44-9def-3592072e344b)

**Comment by ammaree** [comment]
Background:
===========
DHO924S Hardware 8, was running 00.01.04
All features bought up front and was working, never before been patched
Bought the Extended GUI v0.7.1, followed instructions in readme.txt to the letter

Working from MacBook Pro M4, Tahoe 26.1 with VMWare Fusion & Windows 11.
Using Microsoft PowerShell v7.5.4.0

Android Debug Bridge version 1.0.41
Version 36.0.0-13206524
Installed as C:\Users\andre\PlatformTools\adb.exe
Running on Windows 10.0.26100

I tried for possibly 6 hours to find up-to-date instructions to make a backup, could not find anything.
Tried TotalCommander with the ADB Plugin, no success.
Eventually decided to take a chance and go ahead, result as follows:

PS C:\Users\andre\PlatformTools> ./adb connect 192.168.1.38:55555
connected to 192.168.1.38:55555
PS C:\Users\andre\PlatformTools> ./adb root
restarting adbd as root
PS C:\Users\andre\PlatformTools> ./adb shell am force-stop com.rigol.launcher
PS C:\Users\andre\PlatformTools> ./adb shell am force-stop com.rigol.scope
PS C:\Users\andre\PlatformTools> ./adb shell pm uninstall com.rigol.scope
Success
PS C:\Users\andre\PlatformTools> ./adb shell pm uninstall --user 0 com.rigol.scope
Success
PS C:\Users\andre\PlatformTools> ./adb shell mount -o rw,remount /system
PS C:\Users\andre\PlatformTools> ./adb shell mv /system/framework/services.jar /system/framework/services_jar.orig
PS C:\Users\andre\PlatformTools> ./adb push services.jar /system/framework/
services.jar: 1 file pushed, 0 skipped. 9.0 MB/s (3179392 bytes in 0.338s)
PS C:\Users\andre\PlatformTools> ./adb shell mount -o ro,remount /system
PS C:\Users\andre\PlatformTools> ./adb shell sync
PS C:\Users\andre\PlatformTools> ./adb reboot
PS C:\Users\andre\PlatformTools> ./adb connect 192.168.1.38:55555
already connected to 192.168.1.38:55555
PS C:\Users\andre\PlatformTools> ./adb root
restarting adbd as root
PS C:\Users\andre\PlatformTools> cd ../Vendors/Rigol/Sparrow
PS C:\Users\andre\Vendors\Rigol\Sparrow> /Users/andre/PlatformTools/adb install -g -r "Sparrow.apk"
Performing Streamed Install
Success
PS C:\Users\andre\Vendors\Rigol\Sparrow> /Users/andre/PlatformTools/adb reboot
PS C:\Users\andre\Vendors\Rigol\Sparrow>

At this stage the scope reboot but after ~30min wait still just showing the "RIGOL" logo on a blank background...

Any suggestions?

**Reply by mriscoc** [reply]
> PS C:\Users\andre\PlatformTools> ./adb shell mount -o rw,remount /system
> PS C:\Users\andre\PlatformTools> ./adb shell mv /system/framework/services.jar /system/framework/services_jar.orig
> PS C:\Users\andre\PlatformTools> ./adb push services.jar /system/framework/
> services.jar: 1 file pushed, 0 skipped. 9.0 MB/s (3179392 bytes in 0.338s)

It seems that you forgot to rename/remove the odex file, from https://github.com/mriscoc/RIGOL_DHO800_DHO900_GUI/discussions/8:

```shell
adb connect ###.###.###.###:55555
adb root
adb shell mount -o rw,remount /system
adb shell mv /system/framework/services.jar /system/framework/services_jar.orig
adb shell mv /system/framework/oat/arm64/services.odex /system/framework/oat/arm64/services_odex.orig
```

Also, if you have a system that has previously been patched or if you have found this particular .dex file, you must execute the following:
```shell
adb shell rm /data/dalvik-cache/arm64/system@framework@services.jar@classes.dex
```

**Comment by ammaree** [comment]
Have handed it in at local Rigol agent, might cost me a fair bit but I should get it back working, with warranty intact.

The device had never been modified prior to this week's attempt. When I get it back I suspect it will be in exact same never modified configuration. My question is now...
How do I again do the extended GUI installation but 100% correct and in sequence?
What are the exact commend necessary to complete?

All help appreciated,

André M. Maree

**Reply by mriscoc** [reply]
Hi, good news! I'm glad you managed to solve your issue.

**Reply by mriscoc** [reply]
I have deleted some old posts to keep this thread clean. The current installation steps for the updated Sparrow are in the first post, and the steps for gaining system permissions are in the referenced post. The system permissions step is risky and must be carried out carefully, this is why it's separate from the normal installation process.

**Comment by UltraBlackLinux** [comment]
ok EDIT:

I ran through the procedure and ended up on a blank RIGOL screen.

After much finddling around I noticed
```
$ adb shell pm uninstall com.rigol.scope
Failure [DELETE_FAILED_INTERNAL_ERROR
```
and
```
$ adb install -g Sparrow_Extended_v0.7.1-system.apk
Performing Streamed Install
adb: failed to install Sparrow_Extended_v0.7.1-system.apk: Failure [INSTALL_FAILED_ALREADY_EXISTS: Attempt to re-install com.rigol.scope without first uninstalling.]
```

I then tried the solution for the signature mismatch above and I finally have a scope again. Probably worth a notice ;)

**Reply by mriscoc** [reply]
If you have already patched Android to allow the Sparrow system version you must include the "-r" in your installation command:
```adb
adb install -g -r Sparrow_Extended_v0.7.1-system.apk
```

**Comment by aasayag-hash** [comment]
Hello, I have an 802 with version 1.04. I downloaded the files and followed the steps, but I cannot send the .apk. It keeps giving me a signature error. I have attached the steps. What could it be?

```shell
C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4\app>cd..

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb shell am force-stop com.rigol.launcher

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb shell am force-stop com.rigol.scope

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb shell pm clear com.rigol.scope
Success

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb shell pm uninstall --user 0 com.rigol.scope
Success

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb install -g -r "Sparrow.apk"
Performing Streamed Install
adb.exe: failed to stat Sparrow.apk: No such file or directory

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb install -g -r "Sparrow_Extended_v0.7.1-system.apk"
Performing Streamed Install
adb.exe: failed to install Sparrow_Extended_v0.7.1-system.apk: Failure [INSTALL_FAILED_UPDATE_INCOMPATIBLE: Package com.rigol.scope signatures do not match the previously installed version; ignoring!]


C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb shell
rk3399_rigol:/ # su
rk3399_rigol:/ # mount -o rw,remount /system
em/app/Sparrow/Sparrow.apk /system/app/Sparrow/Sparrow.apk.disabled           <
rk3399_rigol:/ # rm -rf /data/app/com.rigol.scope-1
rk3399_rigol:/ # sync
rk3399_rigol:/ # reboot

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb connect 192.168.1.61:55555
already connected to 192.168.1.61:55555

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb root
restarting adbd as root

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb root
adbd is already running as root

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb connect 192.168.1.61:55555
already connected to 192.168.1.61:55555

C:\Users\lenovo\Downloads\scrcpy-win64-v3.3.4>adb install -g -r "Sparrow.apk"
Performing Streamed Install
adb.exe: failed to install Sparrow.apk: Failure [INSTALL_FAILED_UPDATE_INCOMPATIBLE: Package com.rigol.scope signatures do not match the previously installed version; ignoring!]
```
Edited for better formatting

**Reply by UltraBlackLinux** [reply]
read main post.

**Reply by mriscoc** [reply]
> What could it be?

First read each step in the original post, then you need to restore your backup to start again step by step, I think that you are confusing the system and normal versions, the system one requires a patched Android as specified in the instructions.

**Reply by aasayag-hash** [reply]
Thank you, it worked well. I have two questions:
- Will you update the mask with the new font version that comes with update 1.05? It really looks much better.
- I installed the FFT app FFT.Avg_0.3.3. It installed OK, but it does not appear in the DHOtool or action menu options.

**Reply by mriscoc** [reply]
If you already have the DHOTools, you do not need the fftAvg application.

**Reply by youmax01** [reply]
&gt; If you already have the DHOTools, you do not need the fftAvg application.

OK, but DHO tools does not appear in the DHO action menu. I don't know why.

**Reply by UltraBlackLinux** [reply]
the default DHO action menu includes practically nothing. There's another post about changing its config to display more entries.

You can also just install an android launcher and access it that way. Window splitting is done by long-pressing in the app switcher

**Reply by mriscoc** [reply]
> OK, but DHO tools does not appear in the DHO action menu. I don't know why.

The DHOActions includes some examples, but you need to edit the configuration file to add more buttons: https://github.com/mriscoc/Sparrow-Extended-GUI-for-RIGOL-DHO800_DHO900/discussions/9
